CorsixTH/Lua/entities/machine.lua

```diff
@@ -99,29 +99,50 @@ function Machine:machineUsed(room)
   local threshold = self:getRemainingUses()
   -- Find a queued task for a handyman coming to repair this machine
   local taskIndex = self.hospital:getIndexOfTask(self.tile_x, self.tile_y, "repairing")
-
-  -- Too late it is about to explode
+  -- extinguisher check must be reset after each use
+  self.saved_by_extinguisher = 0
+  -- Room is set to explode
   if threshold < 1 then
-    -- Clean up any task of handyman coming to repair the machine
-    self.hospital:removeHandymanTask(taskIndex, "repairing")
-    -- Blow up the room
-    room:crashRoom()
-    self:setCrashedAnimation()
-    -- No special cursor required when hovering over the crashed room
-    self.hover_cursor = nil
-    -- Clear dynamic info (tracks machine usage which is no longer required)
-    self:clearDynamicInfo()
-    -- Prevent the machine from smoking, it's now just a pile of rubble
-    setSmoke(self, false)
-    -- If we have the window for this machine open, close it
-    local window = self.world.ui:getWindow(UIMachine)
-    if window and window.machine == self then
-      window:close()
+    -- If a fire extinguisher in the room, make explosion chance 50%
+    for object, _ in pairs(room.objects) do
+      if object.object_type.id == "extinguisher" then
+        self.saved_by_extinguisher = math.random(0,1)
+        break
+      end
     end
-    -- Clear the icon showing a handyman is coming to repair the machine
-    self:setRepairing(nil)
-    return true
-  -- Else if urgent repair needed
+    -- Extinguisher failed to save the room
+    if self.saved_by_extinguisher == 0 then
+      print("machine lost!")
+      -- Clean up any task of handyman coming to repair the machine
+      self.hospital:removeHandymanTask(taskIndex, "repairing")
+      -- Blow up the room
+      room:crashRoom()
+      self:setCrashedAnimation()
+      -- No special cursor required when hovering over the crashed room
+      self.hover_cursor = nil
+      -- Clear dynamic info (tracks machine usage which is no longer required)
+      self:clearDynamicInfo()
+      -- Prevent the machine from smoking, it's now just a pile of rubble
+      setSmoke(self, false)
+      -- If we have the window for this machine open, close it
+      local window = self.world.ui:getWindow(UIMachine)
+      if window and window.machine == self then
+        window:close()
+      end
+      -- Clear the icon showing a handyman is coming to repair the machine
+      self:setRepairing(nil)
+      return true
+    else
+      -- Extinguisher saved room, machine sure a handyman has been called
+      print("machine saved!")
+      if taskIndex == -1 then
```

Comment:
Not a fan of repeating this; better to split into two conditions:
```lua
local explode = false
if threshold < 1
  -- check if saved by extinguisher, set explode = true otherwise
end
if explode then
  -- do explode type stuff
elseif threshold < 4
  -- do handyman task priority stuff
elseif theshold < 6
  -- other call stuff
end
```